### PR TITLE
cmake: modules: Update zephyr_code_relocate API to support relocating libraries

### PR DESCRIPTION
**Pull request by danieldegrasse**
~Add zephyr_library_relocate, a helper macro intended to allow entire Zephyr libraries to be relocated to a new memory region.~ 

Update `zephyr_code_relocate` macro API to support the following:
- A comma separated list of files (may be used with CMake generator expressions)
- A CMake list of files (useful for `file(GLOB ...)`) directives
- relocating all sources of CMake targets, such as all sources in a Zephyr library

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

### CODE REVIEW

**Comment by carlocaione** [comment]
Please update docs and samples

File: cmake/modules/extensions.cmake
Comment (tejlmand) [inline cmake/modules/extensions.cmake:1284]: instead of creating a new function for this, which needs additional processing in the toplevel CMakeLists.txt file, I would much prefer extending the existing `zephyr_code_relocate()` function
https://github.com/zephyrproject-rtos/zephyr/blob/04684c2837dd8af952aff85885c3be7131c9ae37/cmake/modules/extensions.cmake#L1302

to accept generator expressions.

That avoids having one more function to maintain, and keeps other code clean.

Users can then do something like:
```
zephyr_code_relocate($<TARGET_PROPERTY:library,SOURCES> location)
```

A genex is not configure time processing dependent, making the code itself more robust (generally speaking).

As a small benefit, it will even allow downstream users to relocate files in a Zephyr library without patching Zephyr itself.

**Review by tejlmand — CHANGES_REQUESTED** [review]
Please consider if the existing function can be extended to support generator expressions instead.

File: cmake/modules/extensions.cmake
Comment (danieldegrasse) [reply cmake/modules/extensions.cmake:1284]: I've been looking into this, and I'm not sure how we can generically support CMake generator expressions in this function. If there's a CMake feature I'm unaware of you think might enable this, please let me know.

The core issue is that CMake generators don't seem to have a method of iterating over all files in the target's source property, so we can't handle relative files without changes to the python script. Even if these changes are made, I still don't know of a method to convert the file names to absolute paths using genex expressions, so the benefit you mentioned of downstream users being able to relocate a Zephyr library won't exist.

Another potentially more robust option would be to use the [TARGET_OBJECTS](https://cmake.org/cmake/help/latest/manual/cmake-generator-expressions.7.html#genex:TARGET_OBJECTS) property directly, since this is what the `gen_relocate_app.py` script will do internally. This would require API changes for the `zephyr_code_relocate` macro though, which I'd prefer to avoid.

Comment (tejlmand) [reply cmake/modules/extensions.cmake:1284]: > The core issue is that CMake generators don't seem to have a method of iterating over all files in the target's source property, so we can't handle relative files without changes to the python script. Even if these changes are made, I still don't know of a method to convert the file names to absolute paths using genex expressions, so the benefit you mentioned of downstream users being able to relocate a Zephyr library won't exist.

Easy.
Add this code to any sample CMakeLists.txt and try to run `ninja kernel_sources`:
```
add_custom_target(kernel_sources COMMAND ${CMAKE_COMMAND} -E echo $<TARGET_PROPERTY:kernel,SOURCE_DIR>/$<JOIN:$<TARGET_PROPERTY:kernel,SOURCES>,$<COMMA>$<TARGET_PROPERTY:kernel,SOURCE_DIR>/> )
```

That will print:
```
/projects/github/ncs/zephyr/kernel/main_weak.c,/projects/github/ncs/zephyr/kernel/banner.c,/projects/github/ncs/zephyr/kernel/device.c,/projects/github/ncs/zephyr/kernel/errno.c,/projects/github/ncs/zephyr/kernel/fatal.c,/projects/github/ncs/zephyr/kernel/init.c,/projects/github/ncs/zephyr/kernel/kheap.c,/projects/github/ncs/zephyr/kernel/mem_slab.c,/projects/github/ncs/zephyr/kernel/thread.c,/projects/github/ncs/zephyr/kernel/version.c,/projects/github/ncs/zephyr/kernel/idle.c,/projects/github/ncs/zephyr/kernel/mailbox.c,/projects/github/ncs/zephyr/kernel/msg_q.c,/projects/github/ncs/zephyr/kernel/mutex.c,/projects/github/ncs/zephyr/kernel/queue.c,/projects/github/ncs/zephyr/kernel/sem.c,/projects/github/ncs/zephyr/kernel/stack.c,/projects/github/ncs/zephyr/kernel/system_work_q.c,/projects/github/ncs/zephyr/kernel/work.c,/projects/github/ncs/zephyr/kernel/sched.c,/projects/github/ncs/zephyr/kernel/condvar.c,/projects/github/ncs/zephyr/kernel/xip.c,/projects/github/ncs/zephyr/kernel/timeout.c,/projects/github/ncs/zephyr/kernel/timer.c,/projects/github/ncs/zephyr/kernel/poll.c,/projects/github/ncs/zephyr/kernel/mempool.c
```

Absolute path to all C files included in the kernel library, separated by a comma.

We probably want to provide an easy to use API for the users for the repeated pattern of including a library.
But if we support any generator expression, then we can easily create an easy to use flag.

For example:
```
zephyr_code_relocate(FILES <list of files or generator expression> LOCATION <location>)
```
but even allow a easy to use argument to the specific library case, which then will do a complex generator expression behind the scenes:
```
zephyr_code_relocate(LIBRARY <lib> LOCATION <location>)
```

> Another potentially more robust option would be to use the [TARGET_OBJECTS](https://cmake.org/cmake/help/latest/manual/cmake-generator-expressions.7.html#genex:TARGET_OBJECTS) property directly

With generic genex support, anything is possible.
The user of the function decide which expressions to pass to the function.

> This would require API changes for the zephyr_code_relocate macro though, which I'd prefer to avoid.

Fair point, but considering this PR and #49454, then I believe it's time to cleanup the API for this function and ensure it can handle more use-cases.

Comment (tejlmand) [reply cmake/modules/extensions.cmake:1284]: @danieldegrasse if you're not in a hurry for this, i will be happy to take a look at improving the function myself when having the time, but you are of course more than welcome to play around with this yourself.

Comment (danieldegrasse) [reply cmake/modules/extensions.cmake]: I've updated the `zephyr_code_relocate()` macro based on your suggestions above. Thanks for the pointers- my knowledge on CMake is a bit lacking (especially regarding generator expressions).

File: doc/kernel/code-relocation.rst
Comment (tejlmand) [inline doc/kernel/code-relocation.rst:101]: as described in other comment, then I believe this example will not work unfortunately, because the python relocation script will expect the files to be comma separated,

But let's wait with adjusting this until the final approach has been decided.

**Review by tejlmand — CHANGES_REQUESTED** [review]
Thanks for doing this.

Great improvement, but a couples of details we need to consider / discuss before final approval.

File: cmake/modules/extensions.cmake
Comment (danieldegrasse) [reply cmake/modules/extensions.cmake:1345]: I think the substitution method is probably best, with some documentation provided to the user. It is unfortunate, but if we don't perform some type of modification on the user input then the other alternative I can see be to change the genex directive in [target_relocation.cmake](https://github.com/zephyrproject-rtos/zephyr/blob/ae9de202220ce51e1ed61157f73d3800f84c6b06/cmake/linker/ld/target_relocation.cmake#L21) to something like the following:

```
-i \"$<JOIN:$<TARGET_PROPERTY:code_data_relocation_target,COMPILE_DEFINITIONS>,$<COMMA>>\"
```
which would require the genex passed to `zephyr_library_relocate` to be modified, otherwise this `JOIN` directive would change it as well. So either way, we are left needing to change user input.

If we wanted to place the responsibility on the caller, we could update the `gen_relocate_app.py` script to print a warning when it detects a relocation directive without `:` present, which would be the failure mode of the script when a genex expression was passed without `,` subsituted

Comment (danieldegrasse) [reply cmake/modules/extensions.cmake:1345]: @tejlmand What are your thoughts here? I'd like to move this PR forwards, as #50792 depends on a feature like this being available. Once we settle on a method here I can make other requested updates

**Comment by danieldegrasse** [comment]
@tejlmand I've taken another look at this PR and updated `zephyr_code_relocate` to accept CMake generator expressions directly. I've also extended the `tests/application_development/code_relocation` to test relocating a library, as well as relocating files using a generator expression. Welcome to any feedback you have for these changes!

File: doc/kernel/code-relocation.rst
Comment (danieldegrasse) [reply doc/kernel/code-relocation.rst:89]: I've removed this note in the latest push, so that regular expressions are no longer documented.

File: tests/application_development/code_relocation/CMakeLists.txt
Comment (danieldegrasse) [reply tests/application_development/code_relocation/CMakeLists.txt]: Fixed in latest push

File: scripts/build/gen_relocate_app.py
Comment (danieldegrasse) [reply scripts/build/gen_relocate_app.py]: I've added a deprecation warning to this script, directing the user to use `file(GLOB)` over regex

File: tests/application_development/code_relocation/test_lib/CMakeLists.txt
Comment (danieldegrasse) [inline tests/application_development/code_relocation/test_lib/CMakeLists.txt:4]: I'm not particularly happy with how this library is declared- the goal is to add a library that has the same properties as an in tree Zephyr library, without needing to add a "dummy" library somewhere outside of the sample. I could also use the same approach Zephyr uses for external drivers to add these files, but felt that was unwieldy for a simple test.

File: tests/application_development/code_relocation/test_lib/test_lib1.c
Comment (danieldegrasse) [inline tests/application_development/code_relocation/test_lib/test_lib1.c:11]: @tari When testing these changes, I noticed that these directives can mask errors in the test (on Cortex-M). Since this directive moves symbols to `data.sram2.var` unconditionally, these variables will still be relocated if `zephyr_code_relocate` is not called on this file, so long as another file in the build is relocated to `sram2`, since then the section name will be defined.

I'm interested to know what other approaches we could use here (particularly if `volatile` would prevent these variables from being moved to RODATA). The other option I can see is creating new RAM sections for each file (SRAM2, SRAM3, SRAM4, etc), so that if the file isn't relocated the required linker section won't be present.

Comment (tari) [reply tests/application_development/code_relocation/test_lib/test_lib1.c:11]: I don't think `volatile` would be reliable; it would just enforce that reads always actually read without asserting anything about writability. Creating different sections for each file seems messy but reasonable to me: this is a test so it seems okay if it needs to do some weird things in order to verify behavior that's somewhat complex.

**Comment by carlescufi** [comment]
@carlocaione can you please take a look?

**Comment by carlescufi** [comment]
@tejlmand please re-review

File: cmake/modules/extensions.cmake
Comment (tejlmand) [inline cmake/modules/extensions.cmake:1357]: exactly why is an escaped space needed here ?
That doesn't seem to have been necessary before.

**Review by tejlmand — DISMISSED** [review]
Approved.
A minor comment, but nothing which should prevent merging as it can be updated in followup if need be.

File: cmake/modules/extensions.cmake
Comment (danieldegrasse) [reply cmake/modules/extensions.cmake:1357]: The PHDR directive was previously paired with the relocation location, as you can see in this diff: https://github.com/zephyrproject-rtos/zephyr/pull/50791/files#diff-7087c4d7b0a6595b501520d5eb10ed5ecb01b3540831e3a141a6a662632ac940R20. This directive makes sure that the PHDR is still paired with the code location, so it can be parsed by `gen_relocate_app.py`. `gen_relocate_app.py` treats the PHDR as being paired with the relocation location, so this line preserves that behavior with the new API. An alternative here would be to integrate PHDR as a parameter like NOCOPY is used.

**Review by tejlmand — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]

**Pull request merged**
